File: android/app/src/main/java/com/mccontroller/net/Protocol.kt
package com.mccontroller.net

/**
 * Wire protocol constants. Mirror of the PC server's Protocol.cs.
 * Authoritative spec: docs/protocol.md
 */
object Protocol {
    const val VERSION: Byte = 1
    const val DEFAULT_PORT = 34555

    object MsgType {
        const val HELLO: Byte = 0x01
        const val HELLO_ACK: Byte = 0x02
        const val STATE_CHANGE: Byte = 0x03   // S->C: mode 0/1/2
        const val JOYSTICK: Byte = 0x10
        const val LOOK_DELTA_TCP: Byte = 0x11
        const val LOOK_DELTA_UDP: Byte = 0x11
        const val BUTTON: Byte = 0x20
        const val PING: Byte = 0xF0.toByte()
        const val PONG: Byte = 0xF1.toByte()
        const val PROBE: Byte = 0xFE.toByte()       // C->S, session-less reachability check
        const val PROBE_ACK: Byte = 0xFF.toByte()   // S->C, response to PROBE
    }

    object ProbeStatus {
        const val ALIVE: Byte = 0x00
        const val BUSY: Byte = 0x01
        const val PROTOCOL_INCOMPATIBLE: Byte = 0x02
    }

    object HelloAckStatus {
        const val OK: Byte = 0
        const val PROTOCOL_MISMATCH: Byte = 1
        const val SERVER_BUSY: Byte = 2
    }

    object ButtonId {
        const val MOUSE_LEFT: Byte = 0x01
        const val MOUSE_RIGHT: Byte = 0x02
        const val JUMP: Byte = 0x10
        const val SNEAK: Byte = 0x11
        const val SPRINT: Byte = 0x12
        const val INVENTORY: Byte = 0x20
        const val DROP: Byte = 0x21
        const val SWAP_HAND: Byte = 0x22
        const val ESC: Byte = 0x30
        const val HOTBAR_1: Byte = 0x40
        const val HOTBAR_2: Byte = 0x41
        const val HOTBAR_3: Byte = 0x42
        const val HOTBAR_4: Byte = 0x43
        const val HOTBAR_5: Byte = 0x44
        const val HOTBAR_6: Byte = 0x45
        const val HOTBAR_7: Byte = 0x46
        const val HOTBAR_8: Byte = 0x47
        const val HOTBAR_9: Byte = 0x48
    }
}

File: ios/McController/Net/Protocol.swift
import Foundation

/// Wire protocol constants. Mirror of the PC server's Protocol.cs and the
/// Android client's Protocol.kt.
/// Authoritative spec: docs/protocol.md
enum Protocol {
    static let version: UInt8 = 1
    static let defaultPort: UInt16 = 34555
    static let discoveryPort: UInt16 = 34556

    /// All wire camera-delta values are tenths-of-pixel. The client multiplies
    /// its finger-pixel delta by this before encoding; the server divides.
    static let subpixelScale: CGFloat = 10

    /// Joystick fixed-point: actual_value × 10000 is sent as int16.
    static let joystickScale: Float = 10000

    enum MsgType {
        static let hello: UInt8       = 0x01
        static let helloAck: UInt8    = 0x02
        static let stateChange: UInt8 = 0x03
        static let joystick: UInt8    = 0x10
        static let lookDelta: UInt8   = 0x11   // both TCP and UDP carry this id
        static let button: UInt8      = 0x20
        static let ping: UInt8        = 0xF0
        static let pong: UInt8        = 0xF1
        static let probe: UInt8       = 0xFE
        static let probeAck: UInt8    = 0xFF
    }

    enum ProbeStatus: UInt8 {
        case alive  = 0x00
        case busy   = 0x01
        case incompatible = 0x02
    }

    enum HelloAckStatus: UInt8 {
        case ok               = 0
        case protocolMismatch = 1
        case serverBusy       = 2
    }

    /// PC binding IDs. Mirror of `Protocol.kt` and PC `ButtonId`.
    enum ButtonId: UInt8, CaseIterable {
        case mouseLeft  = 0x01
        case mouseRight = 0x02
        case jump       = 0x10
        case sneak      = 0x11
        case sprint     = 0x12
        case inventory  = 0x20
        case drop       = 0x21
        case swapHand   = 0x22
        case esc        = 0x30
        case hotbar1    = 0x40
        case hotbar2    = 0x41
        case hotbar3    = 0x42
        case hotbar4    = 0x43
        case hotbar5    = 0x44
        case hotbar6    = 0x45
        case hotbar7    = 0x46
        case hotbar8    = 0x47
        case hotbar9    = 0x48

        static func hotbar(_ slot: Int) -> ButtonId {
            precondition((0...8).contains(slot))
            return ButtonId(rawValue: UInt8(0x40 + slot))!
        }
    }
}
